modules/storage-page-memory/src/main/java/org/apache/ignite/internal/storage/pagememory/mv/AbstractPageMemoryMvPartitionStorage.java

```diff
@@ -52,39 +71,134 @@ public abstract class AbstractPageMemoryMvPartitionStorage implements MvPartitio
 
     private static final Predicate<HybridTimestamp> ALWAYS_LOAD_VALUE = timestamp -> true;
 
-    private final int partitionId;
-    private final int groupId;
+    protected final int partitionId;
+
+    protected final AbstractPageMemoryTableStorage tableStorage;
+
+    protected final VersionChainTree versionChainTree;
 
-    private final VersionChainTree versionChainTree;
     protected final RowVersionFreeList rowVersionFreeList;
-    private final DataPageReader rowVersionDataPageReader;
+
+    protected final IndexColumnsFreeList indexFreeList;
+
+    protected final IndexMetaTree indexMetaTree;
+
+    protected final DataPageReader rowVersionDataPageReader;
+
+    protected final ConcurrentMap<UUID, HashIndexStorage> indexes = new ConcurrentHashMap<>();
 
     /**
      * Constructor.
      *
      * @param partitionId Partition id.
-     * @param tableView Table configuration.
-     * @param pageMemory Page memory.
+     * @param tableStorage Table storage instance.
      * @param rowVersionFreeList Free list for {@link RowVersion}.
+     * @param indexFreeList Free list fot {@link IndexColumns}.
      * @param versionChainTree Table tree for {@link VersionChain}.
+     * @param indexMetaTree Tree that contains SQL indexes.
      */
     protected AbstractPageMemoryMvPartitionStorage(
             int partitionId,
-            TableView tableView,
-            PageMemory pageMemory,
+            AbstractPageMemoryTableStorage tableStorage,
             RowVersionFreeList rowVersionFreeList,
-            VersionChainTree versionChainTree
+            IndexColumnsFreeList indexFreeList,
+            VersionChainTree versionChainTree,
+            IndexMetaTree indexMetaTree
     ) {
         this.partitionId = partitionId;
+        this.tableStorage = tableStorage;
 
         this.rowVersionFreeList = rowVersionFreeList;
+        this.indexFreeList = indexFreeList;
+
         this.versionChainTree = versionChainTree;
+        this.indexMetaTree = indexMetaTree;
+
+        PageMemory pageMemory = tableStorage.dataRegion().pageMemory();
 
-        groupId = tableView.tableId();
+        int groupId = tableStorage.configuration().value().tableId();
 
         rowVersionDataPageReader = new DataPageReader(pageMemory, groupId, IoStatisticsHolderNoOp.INSTANCE);
     }
 
+    /**
+     * Starts a partition by initializing its internal structures.
+     */
+    public void start() {
```

Comment:
I feel like there are too much index specifics now (not in method start but in storage class itself). 
Probably it will be bloated in a year. Do you think index stuff can be extracted somehow? I don't have any suggestions though

Reply:
Maybe, but I don't know either. Right now it's not that big, so I don't consider it a problem